### DISCUSSION TITLE
brew rust as dependency

### DISCUSSION THREAD

**Discussion by ricardo-fnd** [Homebrew/discussions#461] [Casks]
Good night!

I'm running a script which installs some programs on 5 macs at each time. **links** is one of the programs and when it reaches that program it automatically installs its dependencies, as normal.
One of them is **rust**, which takes a long time to install (stuck at 'make' for aprox~ 90min).
Is this normal? Every other program is doing good, but is **rust** that heavy?

Thank you.

**Comment by gromgit** [comment]
> links is one of the programs and when it reaches that program it automatically installs its dependencies, as normal. One of them is rust

Since you seem to be building from source for some reason, it's more likely that one of the `links` dependencies in turn depends on `rust` at build time.

> Is this normal?

Under certain conditions, yes. To be sure, what's the output of `brew config` and `brew doctor` on any of the machines currently stuck building `rust`?

**Comment by ricardo-fnd** [comment]
brew config output:

> HOMEBREW_VERSION: 2.7.2
> ORIGIN: https://github.com/Homebrew/brew
> HEAD: dad7dc6a1498b80770d98f2d7cd6fb927c300bbb
> Last commit: 18 hours ago
> Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
> Core tap HEAD: 4867b83c6f29ca092f6c5092b8e871b72c853edf
> Core tap last commit: 3 hours ago
> Core tap branch: master
> HOMEBREW_PREFIX: /usr/local
> HOMEBREW_CASK_OPTS: []
> HOMEBREW_MAKE_JOBS: 4
> Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
> CPU: quad-core 64-bit broadwell
> Clang: 10.0 build 1000
> Git: 2.17.2 => /Library/Developer/CommandLineTools/usr/bin/git
> Curl: 7.54.0 => /usr/bin/curl
> macOS: 10.13.6-x86_64
> CLT: 10.1.0.0.1.1539992718
> Xcode: N/A

brew doctor output:

> Please note that these warnings are just used to help the Homebrew maintainers
> with debugging if you file an issue. If everything you use Homebrew for is
> working fine: please don't worry or file an issue; just ignore this. Thanks!
> 
> Warning: You have unlinked kegs in your Cellar.
> Leaving kegs unlinked can lead to build-trouble and cause brews that depend on
> those kegs to fail to run properly once built. Run `brew link` on these:
>   rust
> 
> Warning: You are using macOS 10.13.
> We (and Apple) do not provide support for this old version.
> You will encounter build failures with some formulae.
> Please create pull requests instead of asking for help on Homebrew's GitHub,
> Twitter or any other official channels. You are responsible for resolving
> any issues you experience while you are running this
> old version.

I can see that rust dependency took 94minutes, downloading 14,000 files and a total of 560MB.

**Reply by gromgit** [reply]
> Warning: You are using macOS 10.13.
> We (and Apple) do not provide support for this old version.

This is the root cause. Every time a formula is version-bumped, the Homebrew CI will only build bottles for whatever OS versions are supported at the time of the build.

`links` should've been installed from bottle, since it hasn't seen a new version in a while, but at least one of its dependencies no longer has a High Sierra bottle, and it depends on `rust` to build...which also has no High Sierra bottle.

So you're stuck with waiting for `rust` and other stuff to build from source.

**Reply by ricardo-fnd** [reply]
The thing is that i'm not stuck, it just takes 94minutes in that specific dependency.

**Reply by gromgit** [reply]
As I said:
> So you're stuck with _waiting_

Unless something fails to build, in which case you're _really_ stuck with a non-working `links`.

**Reply by tomck** [reply]
I'm also hitting this, even major linux distros offer backports, is there a way we could maybe have a community CI or something for High Sierra? That's the cutoff for a lot of still-good macs. It feels weird to want people to use dosdude's patcher just to not have to compile everything. Plus, we could argue the environmental standpoint of having many users burning power just to compile something that could have been compiled once...

**Reply by gromgit** [reply]
> is there a way we could maybe have a community CI or something for High Sierra?

There is, kinda. Someone just has to:
1. Create a new public tap for High Sierra.
2. `brew extract` all the wanted formulae to the new tap, at the versions just before the High Sierra bottles disappears from the formulae.
3. Copy the corresponding bottles from the Homebrew Bintray site to a new site (the URLs should be easy enough to figure out from the output of a normal `brew install`).
4. Edit the `bottle` stanza in all the formulae to add a `root` directive to point to the new bottle site.
5. Fix the new formulae to conform to whatever changes are enforced by the current Homebrew software. (There may be none at all, since this tap would not longer be considered "core".)

That's just to get things going. That person would thereafter have to deal with all the request for new versions, bug fixes, etc.

It's not for the faint of heart, but if you can get others to chip in, the task might not be completely overwhelming.
